gh-108512: Add and use new replacements for PySys_GetObject() (#111035)
python/cpython: Doc/c-api/sys.rst

accessible to C code.  They all work with the current interpreter thread's
:mod:`sys` module's dict, which is contained in the internal thread state structure.

.. c:function:: PyObject *PySys_GetAttr(PyObject *name)

>    Return the object *name* from the :mod:`sys` module or ``NULL`` on failure.
>    Set :exc:`RuntimeError` and return ``NULL`` if it does not exist.

Reviewer: RuntimeError is a little bit surprising. Why not AttributeError?

Does the function return a borrowed reference or a strong reference? Please make it explicit.

I suggest:

```suggestion
   Get the attribute *name* of the :mod:`sys` module. Return a :term:`strong reference`.
   Raise :exc:`RuntimeError` and return ``NULL`` if it does not exist.
```

Author: > RuntimeError is a little bit surprising. Why not AttributeError?

Because it was always raised if `PySys_GetObject()` or variants did not find an object. I just moved it inside the function. Do you prefer to repeat code like `PyErr_SetString(PyExc_RuntimeError, "lost sys.stdin");` at caller place?

> Does the function return a borrowed reference or a strong reference? Please make it explicit.

It returns a strong reference like all other C API, unless opposite is specified.

   If the non-existing object should not be treated as a failure, you can use
   :c:func:`PySys_GetOptionalAttr` instead.

   .. versionadded:: 3.13

.. c:function:: PyObject *PySys_GetAttrString(const char *name)

   This is the same as :c:func:`PySys_GetAttr`, but *name* is
   specified as a :c:expr:`const char*` UTF-8 encoded bytes string,
   rather than a :c:expr:`PyObject*`.

   If the non-existing object should not be treated as a failure, you can use
   :c:func:`PySys_GetOptionalAttrString` instead.

   .. versionadded:: 3.13

.. c:function:: int PySys_GetOptionalAttr(PyObject *name, PyObject **result);

   Variant of :c:func:`PySys_GetAttr` which doesn't raise
   exception if the object does not exist.

>    If the object exists, set *\*result* to a new :term:`strong reference`
>    to the object and return ``1``.
>    If the object does not exist, set *\*result* to ``NULL`` and return ``0``,
>    without setting an exception.
>    If other error occurred, set an exception, set *\*result* to ``NULL`` and
>    return ``-1``.

Reviewer: I suggest to use a list and start with the return value, so it's easier to see the 3 cases:

```suggestion
   * Return ``1`` and set *\*result* to a new :term:`strong reference`
     to the object if the attribute exists.
   * Return ``0`` without setting an exception and set *\*result* to ``NULL``
     if the attribute does not exist.
   * Return ``-1``, set an exception and set *\*result* to ``NULL``
     if an error occurred.
```

Author: "Return and then set exception and variable"  looks like a wrong sequence to me. It cannot do anything after returning.

Reviewer: Just say it in the opposite order in this case:

```
Set an exception, set *\*result* to ``NULL``, and return ``-1``, if an error occurred.
```

accessible to C code.  They all work with the current interpreter thread's
:mod:`sys` module's dict, which is contained in the internal thread state structure.

.. c:function:: PyObject *PySys_GetAttr(PyObject *name)

   Get the attribute *name* of the :mod:`sys` module. Return a :term:`strong reference`.

>    Raise :exc:`RuntimeError` and return ``NULL`` if it does not exist.

Reviewer: I would prefer to raise AttributeError.

Author: Then we would need a new pair of functions that raise RuntimeError to use in the CPython code. Or you can add a pair of functions that raise AttributeError later.

Reviewer: > Then we would need a new pair of functions that raise RuntimeError to use in the CPython code.

Would you mind to elaborate why RuntimeError is needed by CPython?

Author: Status quo. And it looks to me it is justified by the purpose of RuntimeError: "Raised when an error is detected that doesn’t fall in any of the other categories."

If you want to change RuntimeError's in the existing code to something other, you should justify that change.

Reviewer: In my implementation https://github.com/python/cpython/pull/129369, `PySys_SetAttr()` raises `RuntimeError` if the `sys` module cannot be retrieved. But it raises `AttributeError` if the attribute doesn't exist.

`PySys_GetObject()` ignores all errors and returns `NULL` on error or if the attribute doesn't exist. So I'm confused, which function using `PySys_GetObject()` currently rely on the `RuntimeError`? Do you have an example?

Author: About a half of functions that use `PySys_GetObject()` or `_PySys_GetAttr()` raise RuntimeError if they return NULL. See my statistics in https://github.com/capi-workgroup/decisions/issues/54. Look at usages of `PySys_GetAttr()` and `PySys_GetAttrString()` in this PR.

Reviewer: The C API Working Group voted for RuntimeError.

>    Raise :exc:`RuntimeError` and return ``NULL`` if it does not exist.

Reviewer: ```suggestion
   Raise :exc:`RuntimeError` and return ``NULL`` if it does not exist or if the :mod:`sys` module cannot be found.
```
